### PR TITLE
opt(torii): global "sync" for pending & range syncing 

### PR DESCRIPTION
**Pull request by Larkooo**
## Summary by CodeRabbit

- **New Features**
	- Introduced a streamlined synchronization method for blockchain transactions, replacing previous segmented processes with a unified approach.
	- Enhanced efficiency in transaction handling through improved logic that directly manages pending transactions within the synchronization process.

- **Bug Fixes**
	- Adjusted method calls in tests to reflect the new synchronization behavior, ensuring that state management functions correctly with the updated logic.

### CODE REVIEW

File: crates/torii/core/src/engine.rs [hunk 1]
```diff
@@ -268,11 +186,11 @@ impl<P: Provider + Sync> Engine<P> {
...
-                                    to + 1
+                                    latest_block_number + 1
                                 }
                             }
 
-                            _ => to + 1,
+                            _ => latest_block_number + 1,
                         }
                     }
                 };
```
Comment (lambda-0x) [inline crates/torii/core/src/engine.rs:196]: afaik all events (apart from events from pending block) contains the block number, so we can assume if its not present then its an event from pending block. so we dont need to call `get_transction_receipt` for every event in pending block, wdyt?

File: crates/torii/core/src/engine.rs [hunk 2]
```diff
@@ -337,6 +256,11 @@ impl<P: Provider + Sync> Engine<P> {
                 self.process_block(block_number, blocks[&block_number]).await?;
                 last_block = block_number;
             }
+
+            // If the transaction has a world event, we update the cursor
+            if has_world_event {
+                pending_block_tx_cursor = Some(transaction_hash);
+            }
```
Comment (lambda-0x) [inline crates/torii/core/src/engine.rs:263]: since transactions only contain tx from `get_events` i think we don't need this check

Comment (Larkooo) [reply crates/torii/core/src/engine.rs:263]: ah yes we do not anymore

File: crates/torii/core/src/engine.rs (hunk 1 above)
Comment (Larkooo) [reply crates/torii/core/src/engine.rs:196]: yes nice catch.

**Comment by lambda-0x** [comment]
implemented similar optimizations along with erc changes in https://github.com/dojoengine/dojo/pull/2330

**Pull request closed without merging**